### ISSUE TITLE
balanceloss实现问题

### ISSUE THREAD

**Issue by VernonSong** [PaddlePaddle/PaddleOCR#186]
这个实现是不是有错误啊，如果要用这种计算方法的话，loss应该是逐点计算的矩阵，而不是一个值

**Comment by MissPenguin** [comment]
两种方式都可以，效果差不多。

**Comment by kasyoukin** [comment]
有错误啊，配置中配置的

    balance_loss: true  
    main_loss_type: DiceLoss

使用DiceLoss的时候ohem是无效的，如果要用ohem，使用bceloss才可以

    if not balance_loss:
        return loss 

    #此时loss是计算出的一个值，shape 为(N,1),后面的计算是无效的，根本无法选出negative_loss的topK，因为netgativa_loss的值都是一样的，此处是无效计算

    positive_loss = positive * loss
    negative_loss = negative * loss
    negative_loss = fluid.layers.reshape(negative_loss, shape=[-1])
    negative_loss, _ = fluid.layers.topk(negative_loss, k=negative_count_int)
    balance_loss = (fluid.layers.reduce_sum(positive_loss) +
                    fluid.layers.reduce_sum(negative_loss)) / (
                        positive_count + negative_count + eps)

**Comment by jjjjohnson** [comment]
@kasyoukin 我也觉得不对: 计算ohem的mask应该在计算Diceloss之前。并且Diceloss用到了ohem的mask的mask。

**Comment by WenmuZhou** [comment]
大家非常细心啊，这个我们最近也发现了，理论上这里应该使用BCE+OHEM，但是对于目前的后处理来说(有实验证明)，DICE能够取得更好的效果，并且recall和precision更加均衡，我们后面也会进行更多的试验来进行对比提升。
